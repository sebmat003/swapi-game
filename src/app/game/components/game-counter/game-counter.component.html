<div class="row">
  <div *ngFor="let index of [0,1]" class="col-6 text-center">
    <h4>Player {{ index + 1 }}</h4>
    <p [attr.data-cy]="'counter-result'">Result: {{ counter[index]}}</p>
  </div>
</div>
