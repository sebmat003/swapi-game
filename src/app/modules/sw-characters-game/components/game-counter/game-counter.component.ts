import {
  ChangeDetectionStrategy,
  ChangeDetectorRef,
  Component,
  Input,
  OnDestroy,
  OnInit,
} from '@angular/core';
import { Observable, Subject, takeUntil } from 'rxjs';

@Component({
  selector: 'app-game-counter',
  templateUrl: './game-counter.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class GameCounterComponent implements OnInit, OnDestroy {
  @Input() roundWinner$!: Observable<number>;
  destroy$ = new Subject<void>();
  counter: [number, number] = [0, 0];

  constructor(private cdr: ChangeDetectorRef) {}

  ngOnInit(): void {
    this.roundWinner$.pipe(takeUntil(this.destroy$)).subscribe((index) => {
      this.counter[index] += 1;
      this.cdr.detectChanges();
    });
  }

  ngOnDestroy(): void {
    this.destroy$.next();
  }
}
