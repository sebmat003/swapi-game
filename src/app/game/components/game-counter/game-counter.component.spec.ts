import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';
import { GameCounterComponent } from "./game-counter.component";
import { MatSelectModule } from "@angular/material/select";
import { BehaviorSubject } from "rxjs";

describe('GameCounterComponent', () => {
  let spectator: Spectator<GameCounterComponent>;
  let roundWinner$ = new BehaviorSubject<number>(0);
  const createComponent = createComponentFactory({
    component: GameCounterComponent,
    imports: [MatSelectModule],
    detectChanges: false
  });

  beforeEach(() => {
    spectator = createComponent();

    spectator.setInput('roundWinner$', roundWinner$)
  })

  it('Should match snapshot', () => {
    spectator.detectChanges();

    expect(spectator.fixture).toMatchSnapshot();
  });

  it('Should add point to second player', () => {
    roundWinner$.next(1);

    expect(spectator.component.counter).toEqual([1, 1]);
  })
});
