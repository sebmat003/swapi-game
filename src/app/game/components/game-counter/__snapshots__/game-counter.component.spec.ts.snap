// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameCounterComponent Should match snapshot 1`] = `
<app-game-counter
  cdr={[Function ViewRef$1]}
  counter={[Function Array]}
  destroy$={[Function Subject]}
  roundWinner$={[Function Observable]}
>
  <div
    class="row"
  >
    <div
      class="col-6 text-center"
    >
      <h4>
        Player 1
      </h4>
      <p>
        Result: 1
      </p>
    </div>
    <div
      class="col-6 text-center"
    >
      <h4>
        Player 2
      </h4>
      <p>
        Result: 0
      </p>
    </div>
  </div>
</app-game-counter>
`;
